### PR TITLE
Intel OpenVINO backend

### PR DESCRIPTION
**Pull request by dkurt**
# Pull Request Template

## Description

Add Intel OpenVINO backend for prediction. Backend optimizes only predictions, not training.

To enable OpenVINO, add `use_openvino=True` when create a model:

```python
# examples/tox21/tox21_tf_progressive.py

model = dc.models.ProgressiveMultitaskClassifier(
    len(tox21_tasks),
    n_features,
    layer_sizes=[1000],
    dropouts=[.25],
    learning_rate=0.001,
    batch_size=50,
    use_openvino=True)
```

Efficiency measurements (not final, working on improvement):
<details>
<summary>benchmark code</summary>

```python
from __future__ import print_function
from __future__ import division
from __future__ import unicode_literals

import os
import shutil
import numpy as np
import argparse
import deepchem as dc
from deepchem.molnet import load_tox21

parser = argparse.ArgumentParser()
parser.add_argument('--use_openvino', action='store_true')
args = parser.parse_args()

# Only for debug!
np.random.seed(123)

# Load Tox21 dataset
n_features = 1024
tox21_tasks, tox21_datasets, transformers = load_tox21()
train_dataset, valid_dataset, test_dataset = tox21_datasets

# Fit models
metric = dc.metrics.Metric(dc.metrics.roc_auc_score, np.mean)

model = dc.models.ProgressiveMultitaskClassifier(
    len(tox21_tasks),
    n_features,
    layer_sizes=[1000],
    dropouts=[.25],
    learning_rate=0.001,
    batch_size=50,
    use_openvino=args.use_openvino)

print("Evaluating model")
import time
for i in range(3):
  start = time.time()
  train_scores = model.evaluate(train_dataset, [metric], transformers)
  print(time.time() - start)
```

</details>

| tox21_tf_progressive | use_openvino=False | use_openvino=True |
|---|---|---|
| [Intel® Core™ i7 8665UE](https://devcloud.intel.com/edge/devices/intel-core-i7-8665ue-cpu/) | 7.75 sec | 6.03 sec (x1.28) |
| [Intel® Xeon® Gold 6258R](https://devcloud.intel.com/edge/devices/intel-xeon-gold-6258r-cpu/) | 1.23 sec | 0.72 sec (x1.7) |

Tested models:
* ProgressiveMultitaskClassifier (tox21)

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.22.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/openvino_model.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+try:
+  import mo_tf
+  from openvino.inference_engine import IECore
+  is_available = True
+except:
+  is_available = False
+  pass
+
+
+class OpenVINOModel:
```
Comment (rbharath) [inline deepchem/models/openvino_model.py:16]: Could you add a docstring here with documentation on OpenVINO and some more information about the class?

File: deepchem/models/openvino_model.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
+import os
```
Comment (rbharath) [inline deepchem/models/openvino_model.py:1]: Could you add some unit tests for the OpenVINO model to make sure it works correctly?

File: deepchem/models/openvino_model.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+  import mo_tf
+  from openvino.inference_engine import IECore
+  is_available = True
+except:
+  is_available = False
+  pass
+
+
+class OpenVINOModel:
+
```
Comment (rbharath) [inline deepchem/models/openvino_model.py:17]: Could you add OpenVINO to the docs/models.rst file so we get generated autodoc?

**Review by rbharath — COMMENTED** [review]
Thanks for the contribution! I have a few preliminary comments below. I'm not very familiar with OpenVINO so some general information about its use cases and more documentation and tests would be very helpful for us to be able to maintain this as a feature.

**Comment by dkurt** [comment]
@rbharath, Hi! Thanks for review! I fully agree with your comments and going to resolve them soon.

File: deepchem/models/openvino_model.py (hunk 3 above)
Comment (dkurt) [reply deepchem/models/openvino_model.py:17]: As for me, OpenVINOModel seems to be extra in the list. It's just a backend which can be enabled for other Keras (Done) or PyTorch models (TBD).

File: deepchem/models/openvino_model.py (hunk 2 above)
Comment (dkurt) [reply deepchem/models/openvino_model.py:1]: Done

File: deepchem/models/openvino_model.py (hunk 1 above)
Comment (dkurt) [reply deepchem/models/openvino_model.py:27]: Done

File: env.cpu.yml [hunk 4]
```diff
@@ -9,3 +9,5 @@ dependencies:
     - torch-scatter
     - torch-sparse
     - torch-spline-conv
+    - experimental-openvino-python==2021.2.1
+    - experimental-openvino-python-mo==0.0.1
```
Comment (dkurt) [inline env.cpu.yml:13]: I will align the versions

**Comment by dkurt** [comment]
Failed test:
```
deepchem/models/tests/test_graph_models.py::test_dtnn_regression_model FAILED [ 44%]
```
Not sure that it's because of the changes in this PR

File: deepchem/models/openvino_model.py (hunk 3 above)
Comment (dkurt) [reply deepchem/models/openvino_model.py:32]: Added Torch models inference. @rbharath, May I ask to review once again and guide if is there something else good to add?

File: deepchem/models/openvino_model.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+    # Import for pip package
+    from model_optimizer import mo_tf
+  except:
+    # Import for OpenVINO distribution
+    import mo_tf
+  from openvino.inference_engine import IECore, StatusCode
+  is_available = True
+except:
+  is_available = False
+  pass
```
Comment (rbharath) [inline deepchem/models/openvino_model.py:23]: You don't need this extra pass statement

File: deepchem/models/openvino_model.py [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+"""
```
Comment (rbharath) [inline deepchem/models/openvino_model.py:24]: Can you make this docstring the class docstring for `OpenVINOModel`? That way the autodoc can pick it up

File: deepchem/models/openvino_model.py [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+
+  def __init__(self, model, model_dir, batch_size):
+    self.ie = IECore() if self.is_available() else None
+    self.exec_net = None
+    self.model = model
+    self.model_dir = model_dir
+    self.batch_size = batch_size
+    self.outputs = []
+
+  """
```
Comment (rbharath) [inline deepchem/models/openvino_model.py:41]: Can you move this docstring so it's the docstring for the `_read_torch_model` method? We follow numpy conventions (https://deepchem.readthedocs.io/en/latest/development_guide/coding.html#docstrings)

File: deepchem/models/openvino_model.py [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+"""
+Class which wraps Intel OpenVINO toolkit for deep learning inference.
+To enable optimization, pass `use_openvino=True` flag when create a model.
+Read more at https://software.intel.com/content/www/us/en/develop/tools/openvino-toolkit.html
+"""
+
+
+class OpenVINOModel:
+
+  def __init__(self, model, model_dir, batch_size):
```
Comment (rbharath) [inline deepchem/models/openvino_model.py:33]: Can you add type signatures here?

File: deepchem/models/openvino_model.py [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+    import mo_tf
+  from openvino.inference_engine import IECore, StatusCode
+  is_available = True
+except:
+  is_available = False
+  pass
+"""
+Class which wraps Intel OpenVINO toolkit for deep learning inference.
+To enable optimization, pass `use_openvino=True` flag when create a model.
+Read more at https://software.intel.com/content/www/us/en/develop/tools/openvino-toolkit.html
```
Comment (rbharath) [inline deepchem/models/openvino_model.py:27]: It would be useful to expand this out further. When would I want to use this class? Could you add an Examples section to the docstring with an usage example for the class?

File: deepchem/models/openvino_model.py [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+    self.outputs = []
+
+  """
+  Prepare PyTorch model for OpenVINO inference:
+    1. Convert into ONNX format
+    2. Read network from memory
+  NOTE: We do not load model in __init__ method because of training.
+  """
+
+  def _read_torch_model(self, generator):
```
Comment (rbharath) [inline deepchem/models/openvino_model.py:48]: Can you add a type signature?

File: deepchem/models/openvino_model.py [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+
+    buf = io.BytesIO()
+    inp = torch.randn(inp_shape)
+    torch.onnx.export(self.model, inp, buf, opset_version=11)
+
+    # Import network from memory buffer
+    return self.ie.read_network(buf.getvalue(), b'', init_from_buffer=True), \
+           generator
+
+  """
```
Comment (rbharath) [inline deepchem/models/openvino_model.py:67]: Can you move this docstring so it's the docstring for `_read_tf_model`?

File: deepchem/models/openvino_model.py [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+            '--output_dir', self.model_dir
+        ],
+        check=True)
+    os.remove(pb_model_path)
+
+    return self.ie.read_network(
+        os.path.join(self.model_dir, 'model.xml'),
+        os.path.join(self.model_dir, 'model.bin'))
+
+  def _load_model(self, generator, keras_model, torch_model):
```
Comment (rbharath) [inline deepchem/models/openvino_model.py:108]: Can you add type signatures and a docstring here?

File: deepchem/models/openvino_model.py [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+
+    # Load network to the device
+    self.exec_net = self.ie.load_network(
+        net,
+        'CPU',
+        config={'CPU_THROUGHPUT_STREAMS': 'CPU_THROUGHPUT_AUTO'},
+        num_requests=0)
+    return generator
+
+  """
```
Comment (rbharath) [inline deepchem/models/openvino_model.py:123]: Move this docstring to be the docstring for `__call__` below

File: deepchem/models/openvino_model.py [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+  OpenVINO can process data asynchronously.
+  Initialize an iterator for data generator and get outputs by readiness.
+  """
+
+  def __call__(self, generator, keras_model=None, torch_model=None):
```
Comment (rbharath) [inline deepchem/models/openvino_model.py:128]: Can you add type signatures here?

File: deepchem/models/openvino_model.py [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+          self.outputs[out_id] = output[:last_batch_size]
+        else:
+          self.outputs[out_id] = output
+
+    return self, generator
+
+  def __next__(self):
+    return self.outputs.pop(0)
+
+  """
```
Comment (rbharath) [inline deepchem/models/openvino_model.py:201]: Can you move this docstring into `is_available`?

File: env.cpu.yml [hunk 16]
```diff
@@ -9,3 +9,5 @@ dependencies:
     - torch-scatter
     - torch-sparse
     - torch-spline-conv
+    - experimental-openvino-python==2021.2.3
```
Comment (rbharath) [inline env.cpu.yml:12]: How unstable is openvino right now? I'm a little concerned about adding an experimental package right now (our maintenance overhead is a little heavy already).

I'd be OK with this if you're able to commit to maintaining this feature :)

File: deepchem/models/openvino_model.py (hunk 3 above)
Comment (rbharath) [reply deepchem/models/openvino_model.py:32]: Ah that makes sense. Should `OpenVINOModel` be a private class then? If we don't ever expect users to interact with it that might make sense

**Review by rbharath — COMMENTED** [review]
Looks better! Thanks for adding in some unit tests :). I've done a more detailed review pass now

As a first comment, I think we need some more work on the docstrings to match the rest of the codebase.

A second comment is whether we should consider making `OpenVINOModel` a private class.

My other major comment is about maintainability. If OpenVINO is experimental, I want to make sure that we can maintain it. If you have bandwidth to commit to maintaining I think we can make this work :)

Also, tagging in @peastman @nd-02110114 who might be interested to follow along. Please feel free to chime in if you have thoughts!

File: env.cpu.yml (hunk 16 above)
Comment (dkurt) [reply env.cpu.yml:12]: Hi! Actually, it's just about the package. OpenVINO is stable and officially released under `https://pypi.org/project/openvino/`. However it's not compatible with `manylinux` policy and requires extra setup steps (PATH environment variable). So I created a [repo](https://github.com/dkurt/openvino-python-manylinux/) which builds the wheels for the same OpenVINO version but using `manylinux` toolchain and wider Python versions matrix.

File: deepchem/models/openvino_model.py (hunk 9 above)
Comment (dkurt) [reply deepchem/models/openvino_model.py:27]: Added Examples, thanks!

File: deepchem/models/openvino_model.py (hunk 3 above)
Comment (dkurt) [reply deepchem/models/openvino_model.py]: Good point! Moved class to `deepchem/utils`

**Comment by nd-02110114** [comment]
> My other major comment is about maintainability. 

I'm also concerned about this point. 
I seems that OpenVINO is an optimized engine for computer vision such as image processing like CNN. However, most of DNN models for material science rarely treat with image processing and I don't think our users will gain much benefit from it. Also, in the field of chemistry, the inference performance is rarely required as severely as in image recognition or object recognition. 

On the other hand, this PR depends on your published PyPI package and it is personalized. 
So, I seem it is hard for us to maintain.

**Comment by dkurt** [comment]
@nd-02110114, Thanks for feedback!

Despite that OpenVINO was initially designed for computer vision tasks, it's a universal engine and from PR's description you can see that there is a benefit even for networks without convolutions at all (x1.7 improvement for tox21_tf_progressive which consists of GEMM layers only).

> On the other hand, this PR depends on your published PyPI package and it is personalized.
So, I seem it is hard and not motivated for us to maintain.

I see you point. I'll try to switch to an official package.

**Comment by nd-02110114** [comment]
@dkurt I'm sorry for a late response 🙇‍♂️ 

> Despite that OpenVINO was initially designed for computer vision tasks, it's a universal engine and from PR's description you can see that there is a benefit even for networks without convolutions at all (x1.7 improvement for tox21_tf_progressive which consists of GEMM layers only).

I understood the inference performance for tox21_tf_progressive improved. However, as I mentioned, our users like chemists rarely face the situation that requires such inference performance improvements. So, I seem that most of our users don't feel benchmark improvments as a benefit. Generally, our users require  train or preprocess performance improvements rather than inference. When making an inference, the bottleneck of the performance is mainly the preprocess in my experience. If you know the good usecases of OpenVINO in the area of chemistry,  I want to know.

**Comment by rbharath** [comment]
One possible application that comes to mind is if a user wants to run inference against a large chemical or materials library. For example, users sometimes want to run a graph conv model against a large database of compounds like Enamine REAL (~1 billion compounds). In that case, it might be useful to have inference speedups. Could OpenVINO help for this case? If so a small benchmark might really help establish the advantage.

@dkurt Seconding @nd-02110114 that it would be great to hear about any other use cases you have in mind :).

**Comment by dkurt** [comment]
@nd-02110114, @rbharath, Thanks for comments! You're absolutely right that the best benefit of using such kind of optimizations is large volume data.

> For example, users sometimes want to run a graph conv model against a large database of compounds like Enamine REAL (~1 billion compounds)

May I ask to point to a [model](https://github.com/deepchem/deepchem/tree/master/deepchem/models) so we can do benchmarking?

File: env.cpu.yml (hunk 16 above)
Comment (dkurt) [reply env.cpu.yml]: In the latest commit, switched to official OpenVINO pip distribution (tests passed).
There is still one source package which is installed from GitHub repo: https://github.com/dkurt/openvino/tree/mo_pip/model-optimizer (PR https://github.com/openvinotoolkit/openvino/pull/3123)

**Comment by rbharath** [comment]
My apologies for the slow response here! This fell behind due to the DeepChem 2.4.0 release work.

As a suggestion for a benchmark model, could you try running `dc.models.GraphConvModel` with the OpenVino backend to see the speed improvements? You could use any collection of molecules for this, but the ZINC15 dataset is available in MoleculeNet (and is quite large) so might be a good scale benchmark

**Comment by rbharath** [comment]
Closing this old PR for cleanup
